addons/hr_holidays/models/hr_leave.py

```diff
@@ -784,21 +784,47 @@ def _check_validity(self):
             for leave in leaves:
                 employees |= leave._get_employees_from_holiday_type()
             leave_data = leave_type.get_allocation_data(employees, date_from)
+            
+            # Check if negative allocations are allowed
             if leave_type.allows_negative:
                 max_excess = leave_type.max_allowed_negative
                 for employee in employees:
                     if leave_data[employee] and leave_data[employee][0][1]['virtual_remaining_leaves'] < -max_excess:
                         raise ValidationError(_("There is no valid allocation to cover that request."))
                 continue
 
+            # Refined allocation search for relevant allocations only
+            allocation_records = self.env['hr.leave.allocation'].search([
+                ('employee_id', 'in', employees.ids),
+                ('holiday_status_id', '=', leave_type.id),
+                ('state', '=', 'validate'),  # Only consider validated allocations
+                '|',  # Logical OR to handle cases with open-ended allocation dates
+                ('date_to', '=', False),
+                ('date_to', '>=', min(leave.request_date_to for leave in leaves)),
+                '|',
+                ('date_from', '=', False),
+                ('date_from', '<=', max(leave.request_date_from for leave in leaves))
+            ])
+            # Check for any valid allocation covering the leave request period
+            valid_allocation = any(
+                allocation.date_from <= leave.request_date_from <= leave.request_date_to <= allocation.date_to
+                for leave in leaves
+                for allocation in allocation_records
+                if allocation.date_from and allocation.date_to
+            )
```

Comment:
Hello @MohdAlmosawy 
Thanks for the improvements. My previous concerns are now fixed.
A minor Issue that I found:
An employee has a valid allocation that has date_from set as 1st August and no end_date (allocation does not expire).
The employee requests a time off during August. The ValidationError is thrown because the only valid leave is filtered out due to missing a date_to.
Maybe change it to something like the code below. Add "defaults" to the date comparison then there is no need to only check the allocations that have date_from and date_to set. 
```suggestion
            # Check for any valid allocation covering the leave request period
            valid_allocation = any(
                allocation.date_from or datetime.min <= leave.request_date_from <= leave.request_date_to <= allocation.date_to or datetime.max
                for leave in leaves
                for allocation in allocation_records
            )
```